torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -461,6 +465,31 @@ def __init__(
             if param not in ignored_params and not isinstance(param, FlatParameter):
                 params.append(param)
 
+        # At this point we are wrapping an atomic FSDP unit. Check module shapes
+        # and broadcast module from rank 0 to ensure everything starts off with
+        # the same parameters.
+        # Hack: move to compute device and back if it is CPU
+        prev_device = next(module.parameters()).device
+        was_cpu = (prev_device == torch.device("cpu"))
+        if was_cpu:
+            warnings.warn(
+                f"Module is input on CPU, we are moving to to {self.compute_device} to perform parameter verification and will move it back after. In the future please specify device_ids which will explicitly tell FSDP to place the module on a particular device."
+            )
+            module = module.to(self.compute_device)
+        _verify_param_shape_across_processes(
+            process_group=self.process_group,
+            tensors=params,  # TODO: FSDP + DDP to verify buffers as well
```

Comment:
My understanding is that shared parameters will be verified and sync-ed multiple times, but there is not an easy way around this. Is this correct?

Reply:
Yeah, hopefully there are not too many shared params where it becomes a perf issue.